src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,233 @@
-      @processes << process if process.is_a? Process
-      unless @npm_process
-        create_npm_process
-        @npm_process = true
+    private def run_build_command(command, commands)
+      check_directories
+      next_server_commands_range = (1...commands.size)
+      info "Building project #{Helpers.settings.name.colorize(:light_cyan)}"
+      spawn do
+        error_io = IO::Memory.new
+        process = Helpers.run(command, error: error_io)
+        PROCESSES << {process, "server"}
+        loop do
+          if process.terminated?
+            handle_terminaded_process(process, error_io)
+            break
+          end
+          sleep 1
+        end
       end
     end
 
-    private def build_app_process
-      log "Building project #{project_name}..."
-      Amber::CLI::Helpers.run(@build_command)
+    private def handle_terminaded_process(process, error_io)
+      exit_status = process.wait.exit_status
+      if error_io.empty?
+        if exit_status.zero?
+          if @watch_running
+            kill_processes("server")
+          else
+            notify_next_processes
+          end
+          next_server_commands_range.each { @wait_build.send true }
+        else
+          next_server_commands_range.each { @wait_build.send false }
+        end
+      else
+        handle_error(error_io.to_s)
+        next_server_commands_range.each { @wait_build.send false }
+      end
     end
 
-    private def create_watch_process
-      log "Starting #{project_name}..."
-      Amber::CLI::Helpers.run(@run_command, wait: false, shell: false)
+    private def notify_next_processes
+      notify_counter = @notify_counter_channel.receive
+      notify_counter.times { @notify_channel.send nil }
+      @notify_counter_channel.send 0
     end
 
-    private def create_npm_process
-      node_log "Installing dependencies..."
-      Amber::CLI::Helpers.run("npm install --loglevel=error && npm run watch", wait: false)
-      node_log "Watching public directory"
+    private def run_command(command, key)
+      if key == "server"
+        spawn do
+          build_sucess? = @wait_build.receive
+          if build_sucess?
+            error_io = IO::Memory.new
+            process = Helpers.run(command, error: error_io)
+            PROCESSES << {process, "server"}
+            loop do
+              if process.terminated?
+                unless error_io.empty?
+                  handle_error(error_io.to_s)
+                end
+                break
+              end
+              sleep 1
+            end
+          end
+        end
+      else
+        process = Helpers.run(command)
+        PROCESSES << {process, key}
+      end
+    end
+
+    private def kill_processes(key = nil)
+      PROCESSES.each do |process, owner|
+        if process.terminated?
+          PROCESSES.delete(process)
+        elsif owner == key || key.nil?
+          process.kill
+        end
+      end
+    end
+
+    private def error_server(error_output)
```

Comment:
The body of this method should be moved to a ErrorServer class/module

Reply:
I agree, let me split this file :+1: